includes/profile.inc

```diff
@@ -449,3 +459,368 @@ function cm_tools_update_missing_default_field_values_insert_with_languages(arra
+ *       $row['title'] = str_replace('foobar', '', $row['title']);
+ *     }
+ *   );
+ * }
+ * @endcode
+ *
+ * @param array $sandbox
+ *   The sandbox passed to the drupal update hook that this function is called
+ *   from.
+ * @param array|callable $tables
+ *   Either an array of table entries or a callback that returns the next table
+ *   entry on demand. Each entry is an array where the 0th element is the table
+ *   name and the 1st element is an array of the primary key columns.
+ * @param callable $query_alter_callback
+ *   A callback that receives the index query for returning rows from a field
+ *   table. Conditions may be added to it to filter down the list of rows
+ *   processed. It receives these arguments:
+ *   - $query - The DB select query object,
+ *   - $table_name - The name of the table being queried.
+ * @param callable $work_callback
+ *   The callback to apply to each row of each database table. It is called with
+ *   these arguments:
+ *   - $row - Associative array of column name => values for the row. Alter by
+ *            reference,
+ *   - $table_name - The name of the table being queried.
+ *   The $work_callback can return one of these values:
+ *   - NULL (default): Updates the row if a column was changed. Changes to
+ *                     columns used in the primary key are ignored.
+ *   - TRUE: Forces the row to be updated, even if nothing changed,
+ *   - FALSE: Skips row update, even if something changed.
+ * @param int $rows_per_pass
+ *   The number of table rows to load and process in each batch iteration.
+ *
+ * @return string
+ *   A status message suitable for the return value of your update hook.
+ */
+function cm_tools_tables_walk(array &$sandbox, $tables, callable $query_alter_callback, callable $work_callback, $rows_per_pass = 50) {
+/*
+ *              INITIALISE
+ *                   │
+ *                   │  ┌───────────────────────────────────┐
+ *          move to next table     ┌─────────────────────┐  │
+ *                   │  │          │      Next chunk     │  │
+ *           ┌───────▼──▼─────┐    │  ┌────────────────┐ │  │
+ *           │ GATHER ROWS    │    └──► PROCESS ROWS   ├─┘  │
+ *           │ - <table>      ├───────► - <table>      ├────┘
+ *           │                │ First │ - <rows chunk> │
+ *           │                │ chunk └────────────────┘
+ *           └───────┬────────┘
+ *                   │ No tables
+ *                   │  left
+ *                   ▼
+ *                  END
+ */
+  // Convenience reference.
+  if (!isset($sandbox['_cm_tools_tables_walk'])) {
+    $sandbox['_cm_tools_tables_walk'] = array();
+  }
+  $s = &$sandbox['_cm_tools_tables_walk'];
+
+  $status = array();
+
+  if (!is_callable($tables)) {
+    $move_to_next_table = function () use (&$s) {
+      return array_shift($s['tables']);
+    };
+  }
+  else {
+    $move_to_next_table = $tables;
+  }
+
+  // Initialize state and shared variables.
+  if (empty($s)) {
+    $s = array(
```

Comment:
Does this definitely write back to `$sandbox['_cm_tools_tables_walk']`? I can never remember whether setting the value of a variable, which was a reference, just replaces the reference, or updates the value at the referenced place!

Reply:
It does! Because I ran the code and it works!